### DISCUSSION TITLE
[SDK] Clarification for necessity to replace several functions in ethers

### DISCUSSION THREAD

**Discussion by nooma-42** [zkSync-Community-Hub/zksync-developers#346] [SDKs] [answered]
### Environment

Testnet

### Acknowledgement

- [ ] I understand that issues are expected with the Go SDK as they are being updated.

### Issue Description

Some utils are not implemented or override in zksync-ethers, I wonder if I can just use ethers' implementation.
Specifically, I'd like to clarify 4 functions below

1. [`ethers.utils.getAddress(`](https://docs.ethers.org/v5/api/utils/address/#utils-getAddress)
2. [`ethers.constants.AddressZero`](https://docs.ethers.org/v5/api/utils/constants/#constants-AddressZero)
3. [`ContractReceipt`]()
  zkSync-ethers implement transactionReceipt but not contractReceipt
  ```
  export interface ContractReceipt extends TransactionReceipt {
      events?: Array<Event>;
  }
  ```
4. [`ContractTransaction`](https://github.com/ethers-io/ethers.js/blob/ad5f1c5fc7b73cfb20d9012b669e9dcb9122d244/src.ts/contract/types.ts#L51)
When populating a transaction this type is returned.

### Expected Behavior

N/A

### Code Example

N/A

**Comment by Sykoo00o** [comment]
Has this been resolved ?

**Comment by Delaw258** [comment]
In the context of using zksync-ethers, if certain utilities or functionalities are not implemented or overridden, you can indeed resort to using the corresponding functionalities from the ethers.js library directly. Let's address each of the specific functions you mentioned:

ethers.utils.getAddress():
This function is used to normalize and validate Ethereum addresses. If zksync-ethers doesn't implement or overrides this function, you can safely use ethers.utils.getAddress() directly from the ethers.js library.

ethers.constants.AddressZero:
This constant represents the Ethereum address '0x0000000000000000000000000000000000000000'. If zksync-ethers does not provide an equivalent constant, you can use ethers.constants.AddressZero directly from ethers.js.

ContractReceipt:
zksync-ethers does not implement ContractReceipt as a standalone interface but extends TransactionReceipt. If you need additional functionalities specific to contract receipts, you may need to define your own interface or extend TransactionReceipt accordingly. However, you can still use TransactionReceipt provided by zksync-ethers.

ContractTransaction:
This type represents a transaction that is related to a contract. If zksync-ethers does not provide this type or if it returns a different type, you can use ContractTransaction directly from ethers.js when interacting with contracts.

**Reply by nooma-42** [reply]
you didn't seem to explain the nuance of ContractTransaction. To use ContractTransaction in zksync-ethers, I have to make my own interface, it doesn't really make sense to implement it myself. 
I need to know the reference of this answers

**Comment by FrondEnt** [comment]
Interesting

**Comment by JackHamer09** [comment]
@NOOMA-42 Hey! Could you please provide an explanation of the problem you're facing? From the provided description it is not really clear what the issue is. An example of your code or sharing your repo would be very helpful.

**Reply by nooma-42** [reply]
The exact problem is I cannot find the corresponding implementation of functions above in zksync-ethers. These are all the functions in ethers v5, and I'm unsure if I can directly use these in ethers. 1st and 2nd seems fine to me, but I can't use 3rd and 4th cz there's conflict.

Here's below the code samples, and others are in similar case. The exact codebase I'm working on is [a PR to hyperlane](https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/3114).
```typescript
import { AddressZero } from 'zksync-ethers'

// Error
Module '"zksync-ethers"' has no exported member 'AddressZero'.ts(2305)
```

<img width="676" alt="截圖 2024-02-21 上午9 02 01" src="https://github.com/zkSync-Community-Hub/zksync-developers/assets/35304302/5fc7a63a-a20f-42e7-b793-2666bdf69b52">


**My trial:**
I've looked through zksync-ethers code base, it doesn't export AddressZero from inherited ethers package, thus I'm unsure if that's a legal const in the context of zksync

**Reply by JackHamer09** [reply, marked as answer]
@NOOMA-42 most of simple utilities, such as `getAddress`, or constants like `AddressZero`, can be directly exported from the `ethers` package. The `zksync-ethers` may prove useful for actions specifically related to zkSync (e.g., deposit/withdraw methods) or when interacting with signers and providers.

**Comment by gregmolina5** [comment]
Yes, you can use ethers' implementation for the functions mentioned. Since zksync-ethers is built on top of ethers, you can directly utilize ethers' utils and types if they provide the functionality you need.

**Comment by tiovladelfabs** [comment]
You can use ethers.utils.getAddress() and ethers.constants.AddressZero directly from ethers.js.
For ContractReceipt and ContractTransaction, you may need to handle them separately or extend the existing types as needed.

**Comment by webgolfrelenromero** [comment]
Using ethers' implementation for the specified functions should be feasible in most cases, especially if zksync-ethers does not provide an alternative or if the provided implementation does not meet your requirements. You can directly import and utilize ethers' functions in your project as needed.

**Comment by queskim** [comment]
Yes, you can use ethers' implementation for the utils that are not implemented or overridden in zkSync-ethers. For the specific functions mentioned, you can use ethers.utils.getAddress(), ethers.constants.AddressZero, ethers.ContractReceipt, and ethers.ContractTransaction.
